### ISSUE TITLE
__builtin_memcpy/set_inline are broken with msan

### ISSUE THREAD

**Issue by dvyukov** [llvm/llvm-project#57048]
The following program produces a false msan warning:
```
#include <sanitizer/msan_interface.h>
int main(int argc, char** argv) {
	char buf[SIZE];
	__builtin_memset_inline(buf, argc, SIZE);
 	__msan_check_mem_is_initialized(buf, SIZE);
}
```

```
$ clang memset.c -fsanitize=memory -g -O0 -DSIZE=16 && ./a.out
Uninitialized bytes in __msan_check_mem_is_initialized at offset 0 inside [0x7ffc5694e820, 16)
==2658485==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4a9de5 in main memset.c:6:3
    #1 0x7f9aefcc27fc in __libc_start_main csu/../csu/libc-start.c:332:16
```

The warning is produced with all settings on SIZE/-O.

The following program also produces a false warning:
```
#include <sanitizer/msan_interface.h>
int main(int argc, char** argv) {
	char buf1[SIZE];
	char buf2[SIZE] = {argc};
	__builtin_memcpy_inline(buf1, buf2, SIZE);
 	__msan_check_mem_is_initialized(buf1, SIZE);
}
```

```
$ clang memcpy.c -fsanitize=memory -g -O2 -DSIZE=9 && ./a.out
Uninitialized bytes in __msan_check_mem_is_initialized at offset 0 inside [0x7ffcbd03b2c8, 9)
==2658813==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4a9d37 in main memcpy.c:7:3
    #1 0x7f9f8f9697fc in __libc_start_main csu/../csu/libc-start.c:332:16
```

The warning is produced for all SIZEs with -O0 and for SIZE>8 with -O2.

These programs must not produce a warning.

```
clang version 16.0.0 (https://github.com/llvm/llvm-project.git fa250250b2966613fba7b88575029eb7cf0da66c)
```

@melver @eugenis @vitalybuka @kda @gchatelet

**Comment by dvyukov** [comment]
Another thing to double check: if __builtin_memcpy/set_inline are correctly handled by asan/tsan. I suspect they are not, but it only leads to false negatives so unnoticed.

**Comment by melver** [comment]
While mem*.inline promise to calls, would it be fair to make the following exception (worded for hypothetical documentation):

The intrinsic memcpy.inline promises no calls to external functions as an optimization. But with instrumentation enabled it rather difficult to promise that; on top of that, we know that the MSan runtime exists and provides __msan_memcpy(). Therefore, we assume that if instrumentation is requested it's safe to turn memcpy.inline into a call to __msan_memcpy().
Should this be wrong, such as when implementing memcpy() itself, instrumentation should be disabled with the no_sanitize attribute.

**Comment by dvyukov** [comment]
Sounds reasonable. Asan needs reporting callbacks anyway and tsan checks are not inlinable.

**Comment by melver** [comment]
https://reviews.llvm.org/D131577

**Comment by vitalybuka** [comment]
We also move asan/msan to callbacks even for trivial stores as well, as the binary size is concern

**Comment by gchatelet** [comment]
@sivachandra FYI
